### DISCUSSION TITLE
Unable to use Jaro-Winkler Similarity in comparison_levels

### DISCUSSION THREAD

**Discussion by rajeshpai9** [moj-analytical-services/splink#612] [General]
I am trying to use jaro_winkler_sim function in comparisons.

`settings = {
    "link_type": "dedupe_only",
     "comparisons": 
        [{"output_column_name": "FistName",
         "comparison_levels": [
                       {"sql_condition": "FistName_l IS NULL OR FistName_r IS NULL", "label_for_charts": "Null",
                          "is_null_level": True},
                        {
                         "sql_condition": "jaro_winkler_sim(FistName_l, FistName_r) = 1 ", "label_for_charts": "exact jaro winker match"
                        },
                      ]
        }], `

it is throwing me error 
`RuntimeError: Catalog Error: Scalar Function with name jaro_winkler_sim does not exist!
Did you mean "to_minutes"?
LINE 21:     WHEN JARO_WINKLER_SIM(FistName_l, FistName_r) = 1 THEN 2
    WHEN JARO_WINKLER_SIM(FistName_l, FistName_r) < 0.9 AND JARO_WINKLER_SIM(FistName_l, FistName_r) > 0.7 THEN 0
  END AS gamma_FistName
FROM __splink__df_blocked)`

note: java function is imported properly 
`spark.udf.registerJavaFunction('jaro_winkler_sim','uk.gov.moj.dash.linkage.JaroWinklerSimilarity',types.DoubleType())
`
and I am able to verify that using 
query:
`%sql
select jaro_winkler_sim('man', 'mawn')`
output:
` jaro_winkler_sim(man, mawn) = 0.9333333333333333`

so I need help why splink is not recognising registered udf function.

**Comment by RobinL** [comment]
Hello.  I think you're using Splink version 3, and the DuckDB linker, which doesn't use Spark.

If you try with the SparkLinker at:
```
from splink.spark.spark_linker import SparkLinker
```

things might work.

Could you share your full code?

**Reply by rajeshpai9** [reply]
I am using splink==3.0.0.dev18,
and shortest snippet to replicate the issue

```
import splink
from splink.duckdb.duckdb_linker import DuckDBLinker
from pyspark.sql.types import DoubleType
spark.udf.registerJavaFunction('jaro_winkler_sim','uk.gov.moj.dash.linkage.JaroWinklerSimilarity', DoubleType())

linker = DuckDBLinker(df.toPandas())
linker.compute_number_of_comparisons_generated_by_blocking_rule("jaro_winkler_sim(l.firstname ,r.firstname) > 0.9")
```

**Comment by rajeshpai9** [comment]
Hello, Thank you, 
Yes it looks like issue was with using DuckDBLinker and spark,
I was able to resolve with SparkLinker

```
import splink
from pyspark.sql.types import DoubleType
from splink.spark.spark_linker import SparkLinker

spark.udf.registerJavaFunction('jaro_winkler_sim','uk.gov.moj.dash.linkage.JaroWinklerSimilarity', DoubleType())

linker = SparkLinker(df)
linker.compute_number_of_comparisons_generated_by_blocking_rule("jaro_winkler_sim(l.FistName ,r.FistName) > 0.9")
```
